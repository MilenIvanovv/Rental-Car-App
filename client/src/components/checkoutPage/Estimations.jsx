import React, { Component, Fragment } from 'react';
import { connect } from 'react-redux';
import { Link, Redirect } from 'react-router-dom';
import axios from 'axios';
import PropTypes from 'prop-types';
import { API_ROOT } from '../../constants/constants';
import * as calculate from '../../utils/calculate-rent';
import { toastr } from 'react-redux-toastr';
import { setRentals } from '../../actions/setRentalsAction';
import { setCars } from '../../actions/setCarsAction';
import setRentalCarForm from '../../actions/setRentCarFormActions';

import './checkoutPage.css';

export class Estimations extends Component {

  constructor(props) {
    super(props)

    this.state = {
      estimations: {},
      errors: {},
      isDisabled: false,
      redirect: null,
    }

    this.confirmHanlder = this.confirmHanlder.bind(this);
  }

  componentDidUpdate(prevProps) {
    if (JSON.stringify(prevProps.rentCarForm) !== JSON.stringify(this.props.rentCarForm)) {
      this.estimatePrices();
    }
  }

  estimatePrices() {
    const age = this.props.rentCarForm.age.value;
    const returnDate = this.props.rentCarForm.returnDate.value;
    const { car } = this.props;

    if (car && (age.value === '' || returnDate === '')) {
      return;
    }


    const days = calculate.days(new Date(), new Date(returnDate));
    const pricePerDay = calculate.applyAllToPrice(car.price, days, age);
    const totalPrice = calculate.totalPrice(pricePerDay, days);

    this.setState({
      estimations: {
        days,
        pricePerDay,
        totalPrice,
      },
    });
  }

  async confirmHanlder() {
    const { car } = this.props;

    const {
      firstName, lastName, age, returnDate,
    } = this.props.rentCarForm;

    const client = {
      firstName: firstName.value,
      lastName: lastName.value,
      age: age.value,
    };

    try {
      this.setState({ isDisabled: true })
      await new Promise((res) => setTimeout(res, 1000));
      await axios.post(`${API_ROOT}/rentals`, {
        estimatedDate: new Date(returnDate.value),
        client,
        carId: car.id,
      });
      await this.getCurrentRentals();
      await this.getCars();
      this.props.resetForm();
      this.setState({ redirect: '/current-rentals' });
      toastr.success('Car rented', 'You successfully rented a car!');
    } catch (error) {
      toastr.error('Car renting failed', 'Error occureed while renting a car!');
      console.log(error);
    }
    this.setState({ isDisabled: false })
  }

  async getCurrentRentals() {
    await new Promise((res, rej) => {
      setTimeout(res, 1000);
    })
    // eslint-disable-next-line no-shadow
    const { setRentals } = this.props;
    try {
      const rentals = await axios.get(`${API_ROOT}/rentals`);
      setRentals(rentals.data);
    } catch (error) {
      console.log(error);
    }
  }

  async getCars() {
    const { setCars: dispatchSetCars } = this.props;

    let cars;
    await new Promise((res) => setTimeout(res, 1000));
    try {
      cars = await axios.get(`${API_ROOT}/cars`);
    } catch (error) {
      console.log(error);
    }
    dispatchSetCars(cars.data);
  }

  render() {

    const {
      estimations,
      isDisabled,
      redirect
    } = this.state;

    const {
      isFormValid
    } = this.props;

    if (redirect) {
      return <Redirect to={redirect} />;
    }

    return (
      <div className="card" style={{ width: '18rem' }}>
        <div className="card-body align-card-text">
          <div className="card-text">
            <p className="w-50">
              <span>Days</span>
              <span>Price per day</span>
              <span>Total</span>
            </p>
            <p>
              <span>{estimations.days || 0}</span>
              <span>{`${estimations.pricePerDay || 0} $`}</span>
              <span>{`${estimations.totalPrice || 0} $`}</span>
            </p>
          </div>
          <div className="d-flex justify-content-around">
            <button data="confirm" type="button" className="btn btn-primary" disabled={!isFormValid || isDisabled} onClick={this.confirmHanlder}>
              Confirm
            </button>
            <Link to="/cars">
              <button type="button" className="btn btn-primary">
                Cencel
              </button>
            </Link>
          </div>
        </div>
      </div>
    );
  }
}

Estimations.propTypes = {
  rentCarForm: PropTypes.shape({
    firstName: PropTypes.shape({
      value: PropTypes.string.isRequired,
      error: PropTypes.string.isRequired,
    }),
    lastName: PropTypes.shape({
      value: PropTypes.string.isRequired,
      error: PropTypes.string.isRequired,
    }),
    age: PropTypes.shape({
      value: PropTypes.string.isRequired,
      error: PropTypes.string.isRequired,
    }),
    returnDate: PropTypes.shape({
      value: PropTypes.string.isRequired,
      error: PropTypes.string.isRequired,
    }),
    isFormValid: PropTypes.bool.isRequired,
  }),
  isFormValid: PropTypes.bool.isRequired,
};

Estimations.defaultProps = {
  car: null,
};

const mapStateToProps = (state) => ({
  rentCarForm: state.form,
  isFormValid: state.form.isFormValid,
});

const mapActionsToProps = {
  setRentals,
  setCars,
  resetForm: setRentalCarForm.resetForm
}

export default connect(mapStateToProps, mapActionsToProps)(Estimations);
