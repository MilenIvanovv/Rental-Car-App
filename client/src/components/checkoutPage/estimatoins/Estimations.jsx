import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Link, Redirect } from 'react-router-dom';
import axios from 'axios';
import PropTypes from 'prop-types';
import { toastr } from 'react-redux-toastr';
import { API_ROOT } from '../../../constants/constants';
import * as calculate from '../../../utils/calculate-rent';
import { setRentals } from '../../../actions/setRentalsAction';
import { setCars } from '../../../actions/setCarsAction';
import setRentalCarForm from '../../../actions/setRentCarFormActions';
import { carTypes } from '../../../common/models/prop-types';

import './estimations.css';
import '../checkoutPage.css';

export class Estimations extends Component {
  constructor(props) {
    super(props);

    this.state = {
      estimations: {},
      isDisabled: false,
      redirect: null,
    };

    this.confirmHanlder = this.confirmHanlder.bind(this);
  }

  componentDidUpdate(prevProps) {
    const { rentCarForm } = this.props;

    if (JSON.stringify(prevProps.rentCarForm) !== JSON.stringify(rentCarForm)
      && (rentCarForm.age.value !== '' && rentCarForm.returnDate.value !== null)) {
      this.estimatePrices();
    }
  }

  async getCurrentRentals() {
    // eslint-disable-next-line no-shadow
    const { setRentals } = this.props;
    try {
      const rentals = await axios.get(`${API_ROOT}/rentals`);
      setRentals(rentals.data);
    } catch (error) {
      console.log(error);
    }
  }

  async getCars() {
    const { setCars: dispatchSetCars } = this.props;

    let cars;
    try {
      cars = await axios.get(`${API_ROOT}/cars`);
    } catch (error) {
      console.log(error);
    }
    dispatchSetCars(cars.data);
  }

  validateForm() {
    const { rentCarForm, modifyForm } = this.props;
    const tempForm = JSON.parse(JSON.stringify(rentCarForm));
    let isValid = true;
    Object.values(tempForm).forEach((field) => {
      if (field.error === 'not touched') {
        // eslint-disable-next-line no-param-reassign
        field.error = 'cannot be empty!';
        isValid = false;
      }
    });

    tempForm.isFormValid = isValid;
    modifyForm(tempForm);
    return isValid;
  }

  estimatePrices() {
    const { car, rentCarForm } = this.props;
    const age = rentCarForm.age.value;
    const returnDate = rentCarForm.returnDate.value;

    if (car && (age.value === '' || returnDate === '')) {
      return;
    }

    const days = calculate.days(new Date(), new Date(returnDate));
    const pricePerDay = calculate.applyAllToPrice(car.price, days, age);
    const totalPrice = calculate.totalPrice(pricePerDay, days);

    this.setState({
      estimations: {
        days,
        pricePerDay,
        totalPrice,
      },
    });
  }

  async confirmHanlder() {
    if (!this.validateForm()) {
      return;
    }

    const { car, rentCarForm, resetForm } = this.props;

    const {
      firstName, lastName, age, returnDate,
    } = rentCarForm;

    const client = {
      firstName: firstName.value,
      lastName: lastName.value,
      age: +age.value,
    };

    try {
      this.setState({ isDisabled: true });
      await axios.post(`${API_ROOT}/rentals`, {
        estimatedDate: new Date(returnDate.value),
        client,
        carId: car.id,
      });
      await this.getCurrentRentals();
      await this.getCars();
      resetForm();
      this.setState({ redirect: '/current-rentals' });
      toastr.success('Car rented', 'You successfully rented a car!');
    } catch (error) {
      toastr.error('Car renting failed', 'Error occureed while renting a car!');
      console.log(error);
    }
    this.setState({ isDisabled: false });
  }

  render() {
    const {
      estimations,
      isDisabled,
      redirect,
    } = this.state;

    if (redirect) {
      return <Redirect to={redirect} />;
    }

    return (
      <div className="estimations card">
        <div className="card-body align-card-text">
          <div className="card-text">
            <p className="w-50">
              <span>Days</span>
              <span>Price per day</span>
              <span>Total</span>
            </p>
            <p>
              <span>{estimations.days || 0}</span>
              <span>{`$${estimations.pricePerDay || 0}`}</span>
              <span>{`$${estimations.totalPrice || 0}`}</span>
            </p>
          </div>
          <div className="buttons-container">
            <button data="confirm" type="button" className="btn btn-primary" disabled={isDisabled} onClick={this.confirmHanlder}>
              Confirm
            </button>
            <Link to="/cars">
              <button type="button" className="btn btn-secondary">
                Cencel
              </button>
            </Link>
          </div>
        </div>
      </div>
    );
  }
}

Estimations.propTypes = {
  rentCarForm: PropTypes.shape({
    firstName: PropTypes.shape({
      value: PropTypes.string.isRequired,
      error: PropTypes.string.isRequired,
    }),
    lastName: PropTypes.shape({
      value: PropTypes.string.isRequired,
      error: PropTypes.string.isRequired,
    }),
    age: PropTypes.shape({
      value: PropTypes.string.isRequired,
      error: PropTypes.string.isRequired,
    }),
    returnDate: PropTypes.shape({
      value: PropTypes.string,
      error: PropTypes.string.isRequired,
    }),
    isFormValid: PropTypes.bool.isRequired,
  }).isRequired,
  car: PropTypes.shape(carTypes),
  setRentals: PropTypes.func.isRequired,
  setCars: PropTypes.func.isRequired,
  resetForm: PropTypes.func.isRequired,
  modifyForm: PropTypes.func.isRequired,
};

Estimations.defaultProps = {
  car: null,
};

const mapStateToProps = (state) => ({
  rentCarForm: state.form,
});

const mapActionsToProps = {
  setRentals,
  setCars,
  resetForm: setRentalCarForm.resetForm,
  modifyForm: setRentalCarForm.modifyForm,
};

export default connect(mapStateToProps, mapActionsToProps)(Estimations);
