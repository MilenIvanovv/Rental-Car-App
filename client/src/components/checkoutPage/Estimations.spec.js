import React from "react";
import { create } from "react-test-renderer";
import Estimations from './Estimations';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router';

const mockStore = configureStore([]);

// describe("Estimations component", () => {
//   test("Matches the snapshot", () => {
//     const button = create(<Estimations />);
//     expect(button.toJSON()).toMatchSnapshot();
//   });
// });

describe('My Connected React-Redux Component', () => {
  let store;
  let component;

  beforeEach(() => {

  });

  it('should render with given state from Redux store', () => {
    store = mockStore({
      form: {
        firstName: {
          value: '',
          error: 'not touched'
        },
        lastName: {
          value: '',
          error: 'not touched'
        },
        age: {
          value: '',
          error: 'not touched'
        },
        returnDate: {
          value: '2020-03-18T12:18',
          error: 'not touched'
        },
        isFormValid: false
      }
    });

    component = renderer.create(
      <Provider store={store}>
        <MemoryRouter>
          <Estimations />
        </MemoryRouter>
      </Provider>
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('confirm button should be disabled if form is not valid', () => {

    store = mockStore({
      form: {
        firstName: {
          value: '',
          error: 'not touched'
        },
        lastName: {
          value: '',
          error: 'not touched'
        },
        age: {
          value: '',
          error: 'not touched'
        },
        returnDate: {
          value: '2020-03-18T12:18',
          error: 'not touched'
        },
        isFormValid: false
      }
    });

    component = renderer.create(
      <Provider store={store}>
        <MemoryRouter>
          <Estimations />
        </MemoryRouter>
      </Provider>
    );

    const btn = component.root.findAllByType('button')

    expect(btn[0].props.disabled).toEqual(true);
  });

  it('confirm button should be active if form is not valid', () => {

    store = mockStore({
      form: {
        firstName: {
          value: 'test',
          error: 'not touched'
        },
        lastName: {
          value: 'test',
          error: 'not touched'
        },
        age: {
          value: 'test',
          error: 'not touched'
        },
        returnDate: {
          value: '2020-03-18T12:18',
          error: 'not touched'
        },
        isFormValid: true
      }
    });

    component = renderer.create(
      <Provider store={store}>
        <MemoryRouter>
          <Estimations />
        </MemoryRouter>
      </Provider>
    );

    const btn = component.root.findAllByType('button')

    expect(btn[0].props.disabled).toEqual(false);
  });
});