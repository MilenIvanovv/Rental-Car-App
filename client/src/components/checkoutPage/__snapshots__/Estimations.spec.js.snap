exports[`My Connected React-Redux Component should render with given state from Redux store 1`] = `
<div
  className="card"
  style={
    Object {
      "width": "18rem",
    }
  }>
  <div
    className="card-body align-card-text">
    <div
      className="card-text">
      <p
        className="w-50">
        <span>
          Days
        </span>
        <span>
          Price per day
        </span>
        <span>
          Total
        </span>
      </p>
      <p>
        <span>
          0
        </span>
        <span>
          0 $
        </span>
        <span>
          0 $
        </span>
      </p>
    </div>
    <div
      className="buttons-container">
      <button
        className="btn btn-primary"
        data="confirm"
        disabled={false}
        onClick={[Function]}
        type="button">
        Confirm
      </button>
      <a
        href="/cars"
        onClick={[Function]}>
        <button
          className="btn btn-primary"
          type="button">
          Cencel
        </button>
      </a>
    </div>
  </div>
</div>
`;
